### PR TITLE
Improve handling of set_name! on components

### PR DESCRIPTION
**Pull request by daniel-thom**
CI will fail until https://github.com/NREL-SIIP/InfrastructureSystems.jl/pull/268 is merged and tagged.

The purpose of this PR is block users from changing the name of a component that is attached to a system with `set_name!(component, name)`.

### CODE REVIEW

File: src/base.jl [hunk 1]
```diff
@@ -694,11 +694,7 @@ Throws ArgumentError if the type is not stored.
 """
 # the argument order in this function is un-julian and should be deprecated in 2.0
 function remove_components!(::Type{T}, sys::System) where {T <: Component}
-    components = IS.remove_components!(T, sys.data)
-    for component in components
-        handle_component_removal!(sys, component)
-    end
-    return components
+    return remove_components!(sys, T)
```
Comment (daniel-thom) [inline src/base.jl:697]: This removes redundant code.

File: src/base.jl [hunk 2]
```diff
@@ -1649,12 +1663,12 @@ function handle_component_addition!(sys::System, component::RegulationDevice; kw
 end
 
 function handle_component_addition!(sys::System, component::Branch; kwargs...)
-    handle_component_addition_common!(sys, component)
+    _handle_branch_addition_common!(sys, component)
```
Comment (daniel-thom) [inline src/base.jl:1666]: This was a bad function name.

File: src/base.jl [hunk 3]
```diff
@@ -1740,6 +1759,10 @@ function handle_component_removal!(sys::System, dyn_injector::DynamicInjection)
     @assert found
 end
 
+function _handle_component_removal_common!(component)
+    clear_units!(component)
```
Comment (daniel-thom) [inline src/base.jl:1763]: We were not clearing the units in all removal functions. That should be fixed now.

File: src/descriptors/power_system_structs.json [hunk 4]
```diff
@@ -221,6 +221,7 @@
         {
           "null_value": "init",
           "name": "name",
+          "exclude_setter": true,
```
Comment (daniel-thom) [inline src/descriptors/power_system_structs.json:224]: People usually copy/paste these objects when creating new structs. You all will need to remember to ensure people add this field.

**Review by claytonpbarrows — APPROVED** [review]
This looks good to me. As I understand it, the only downside is that the `exclude_setter: true` entry is critical for every `name` field in the struct definitions.

**Comment by daniel-thom** [comment]
> This looks good to me. As I understand it, the only downside is that the `exclude_setter: true` entry is critical for every `name` field in the struct definitions.

Correct. They will get the current behavior.

**Comment by daniel-thom** [comment]
@claytonpbarrows Note that we'll need to apply these changes to FuelSystems after this gets merged.

**Review by jd-lara — APPROVED** [review]

**Pull request merged**